core/base/inc/Rtypes.h

```diff
@@ -225,44 +225,159 @@ namespace ROOT {
 #ifndef ROOT_TGenericClassInfo
 #include "TGenericClassInfo.h"
 #endif
+#ifndef ROOT_TIsAProxy
+#include "TIsAProxy.h"
+#endif
+
 
 typedef std::atomic<TClass*> atomic_TClass_ptr;
 
+// Parts or ROOT are compiled with no-rtti and cannot use or even see typeid().
+// Hide the following declaration (that is only needed by ClassDefInline) for
+// these cases by #defining R__NO_INLINE_CLASSDEF:
+#ifndef R__NO_INLINE_CLASSDEF
+#include "RStringView.h"
+namespace TClassEdit {
+   void GetNormalizedName(std::string &norm_name,std::string_view name);
+}
+
+namespace ROOT { namespace Internal {
+/// \class TypeNameExtraction
+/// Extracts the fully qualified type name by checking for the name of a
+/// member function as determined by the __PRETTY_FUNCTION__ macro.
+template <class T>
+   struct TypeNameExtraction {
+      static std::string get() {
+#ifdef _MSC_VER // Visual Studio
+# define  __PRETTY_FUNCTION__ __FUNCTION__
+#endif
+         constexpr static const char* funcname = __PRETTY_FUNCTION__;
+         const char* start = strstr(funcname, "TypeNameExtraction<");
+         const char* end = strstr(funcname, ">::get(");
+         if (!start || !end)
+            return "";
+         start += 19; // len of "TypeNameExtraction<"
+         return std::string(start, end - start);
+      }
+   };
+
+template <typename T>
+   class ClassDefGenerateInitInstanceLocalInjector {
+      static atomic_TClass_ptr fgIsA;
+      static std::string fgName;
+      static std::mutex fgMutex;
+   public:
+      static void *New(void *p) { return p ? new(p) T : new T; };
+      static void *NewArray(Long_t nElements, void *p) {
+         return p ? new(p) T[nElements] : new T[nElements]; }
+      static void Delete(void *p) { delete ((T*)p); }
+      static void DeleteArray(void *p) { delete[] ((T*)p); }
+      static void Destruct(void *p) { ((T*)p)->~T();  }
+      static ::ROOT::TGenericClassInfo *GenerateInitInstanceLocal() {
+         static ::TVirtualIsAProxy* isa_proxy = new ::TInstrumentedIsAProxy<T>(0);
+         static ::ROOT::TGenericClassInfo
+            R__instance(T::Class_Name(), T::Class_Version(),
+                        T::DeclFileName(), T::DeclFileLine(),
+                        typeid(T), ROOT::Internal::DefineBehavior((T*)0, (T*)0),
+                        &T::Dictionary, isa_proxy, 0, sizeof(T) );
+         std::lock_guard<std::mutex> guard(fgMutex);
+         R__instance.SetNew(&New);
+         R__instance.SetNewArray(&NewArray);
+         R__instance.SetDelete(&Delete);
+         R__instance.SetDeleteArray(&DeleteArray);
+         R__instance.SetDestructor(&Destruct);
+         return &R__instance;
+      }
+      static void Dictionary() { fgIsA = GenerateInitInstanceLocal()->GetClass(); }
+      static TClass *Class() { if (!fgIsA) Dictionary(); return fgIsA; }
+      static const char* Name() {
+         if (fgName.empty()) {
+            std::lock_guard<std::mutex> guard(fgMutex);
+            if (fgName.empty()) {
```

Comment:
The use of a mutex other that gROOTMutex here guarantees that the code will dead-lock (because TClassEdit::GetNormalizedName can take gROOTMutex).